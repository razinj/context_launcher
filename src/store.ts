import AsyncStorage from '@react-native-async-storage/async-storage'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import {
  createMigrate,
  FLUSH,
  MigrationManifest,
  PAUSE,
  PERSIST,
  PersistConfig,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist'
import createSagaMiddleware from 'redux-saga'
import rootSaga from './rootSaga'
import appsListReducer from './slices/appsList'
import appStateReducer from './slices/appState'
import FavoriteAppsReducer from './slices/favoriteApps'
import pinnedAppsReducer from './slices/pinnedApps'
import preferencesReducer from './slices/preferences'
import recentAppsReducer from './slices/recentApps'

export const rootReducer = combineReducers({
  appsList: appsListReducer,
  recentApps: recentAppsReducer,
  favoriteApps: FavoriteAppsReducer,
  preferences: preferencesReducer,
  pinnedApps: pinnedAppsReducer,
  appState: appStateReducer,
})

const migrations: MigrationManifest = {
  2: (state: any) => {
    return {
      ...state,
      preferences: {
        displayRecentApps: true,
        displayFavoriteApps: true,
      },
    }
  },
  3: (state: any) => {
    return {
      ...state,
      pinnedApps: {
        list: [],
        temporaryAppsConfig: {
          startDate: undefined,
          endDate: undefined,
        },
      },
      preferences: {
        ...state.preferences,
        displayPinnedApps: true,
        displayTemporaryPinnedApps: false,
      },
    }
  },
  4: (state: any) => {
    const localState = { ...state }
    delete localState.appsSearch

    return {
      ...localState,
      appsList: {
        list: [],
      },
      pinnedApps: {
        list: [],
        temporarily: [],
        temporaryAppsConfig: {
          startDate: undefined,
          endDate: undefined,
        },
      },
      appState: {
        displayAllApps: false,
        displaySettings: false,
        menuAppDetails: undefined,
        displayAppMenu: false,
        displaySortableFavoriteApps: false,
        displaySortablePinnedApps: false,
        search: {
          query: undefined,
          result: [],
        },
      },
    }
  },
}

const persistConfig: PersistConfig<any> = {
  key: 'root',
  debug: false,
  version: 4,
  storage: AsyncStorage,
  blacklist: ['appState'],
  migrate: createMigrate(migrations, { debug: false }),
}

const persistedReducer = persistReducer(persistConfig, rootReducer)

const sagaMiddleware = createSagaMiddleware()

export const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }).concat(sagaMiddleware),
})

sagaMiddleware.run(rootSaga)

export const persistor = persistStore(store)

export type RootState = ReturnType<typeof rootReducer>
export type AppDispatch = typeof store.dispatch
