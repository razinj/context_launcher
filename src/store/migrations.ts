import { MigrationManifest } from 'redux-persist'

export const migrations: MigrationManifest = {
  2: (state: any) => {
    return {
      ...state,
      preferences: {
        displayRecentApps: true,
        displayFavoriteApps: true,
      },
    }
  },
  3: (state: any) => {
    return {
      ...state,
      pinnedApps: {
        list: [],
        temporaryAppsConfig: {
          startDate: undefined,
          endDate: undefined,
        },
      },
      preferences: {
        ...state.preferences,
        displayPinnedApps: true,
        displayTemporaryPinnedApps: false,
      },
    }
  },
  4: (state: any) => {
    const localState = { ...state }
    delete localState.appsSearch

    return {
      ...localState,
      appsList: {
        list: [],
      },
      pinnedApps: {
        list: [],
        temporarily: [],
        temporaryAppsConfig: {
          startDate: undefined,
          endDate: undefined,
        },
      },
      appState: {
        displayAllApps: false,
        displaySettings: false,
        menuAppDetails: undefined,
        displayAppMenu: false,
        displaySortableFavoriteApps: false,
        displaySortablePinnedApps: false,
        search: {
          query: undefined,
          result: [],
        },
      },
    }
  },
}
