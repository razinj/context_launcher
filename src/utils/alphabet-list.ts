import { AppDetails } from '../models/app-details'
import { AppLetterIndex } from '../models/list-letter-index'

type LetterMap = {
  [key: string]: number
}

const alphabet = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
]

/**
 * @returns Indexed alphabet letters in an object where the key is the alphabet and the value is the index
 */
export const getLettersMap = (): LetterMap => {
  const letterMap: LetterMap = {}

  alphabet.forEach((letter: string, index: number) => (letterMap[letter] = index + 1))

  return letterMap
}

/**
 * @param value Any string value
 * @returns A letter if the first chareceter of the value is an alphabet, otherwise it returns a '#'
 */
export const getFirstLetter = (value: string): string => {
  const firstChar = value.trim().substring(0, 1).toUpperCase()
  const isValidLetter = getLettersMap()[firstChar]

  if (isValidLetter) return firstChar

  return '#'
}

export const getAppsLetterIndex = (apps: AppDetails[]): AppLetterIndex[] => {
  const appsLetterIndex: AppLetterIndex[] = []
  const treatedLetters: string[] = []

  apps.forEach(({ name }: AppDetails, index: number) => {
    const letter = getFirstLetter(name)

    if (treatedLetters.includes(letter)) return

    appsLetterIndex.push({ letter, index })
    treatedLetters.push(letter)
  })

  return appsLetterIndex
}
