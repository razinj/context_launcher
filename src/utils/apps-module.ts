// Native modules
import AppsModule from '../native-modules/AppsModule'

export const launchApp = (packageName: string) => AppsModule.launchApplication(packageName)

export const showAppDetails = (packageName: string) => AppsModule.showApplicationDetails(packageName)

export const requestAppUninstall = (packageName: string) => AppsModule.requestApplicationUninstall(packageName)
