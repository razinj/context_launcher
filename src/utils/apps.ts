import { APP_ITEM_HEIGHT } from '../constants'
import { AppDetails } from '../models/app-details'
import { PinnedApp } from '../models/pinned-app'

export const getAppsByLabel = (apps: AppDetails[], query: string): AppDetails[] => {
  return apps.filter((app: AppDetails) => app.name.match(new RegExp(query, 'gi')))
}

export const getAppIndex = (apps: AppDetails[], appPackageName: string): number => {
  return apps.findIndex(({ packageName }: AppDetails) => packageName === appPackageName)
}

export const getApp = (apps: AppDetails[], appPackageName: string): AppDetails | PinnedApp | undefined => {
  return apps.find(({ packageName }: AppDetails) => packageName === appPackageName)
}

export const getListKey = ({ packageName }: AppDetails) => packageName

export const getListItemLayout = (_data: unknown, index: number) => ({
  length: APP_ITEM_HEIGHT,
  offset: APP_ITEM_HEIGHT * index,
  index,
})
