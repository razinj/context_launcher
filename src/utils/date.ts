export const getTimeFromDate = (stringDate?: string): string => {
  if (!stringDate) return 'unset'

  const date = new Date(stringDate)

  const singleDigitHours = `${date.getHours()}`.length == 1
  const singleDigitMinutes = `${date.getMinutes()}`.length == 1

  const hours = singleDigitHours ? `0${date.getHours()}` : date.getHours()
  const minutes = singleDigitMinutes ? `0${date.getMinutes()}` : date.getMinutes()

  return `${hours}:${minutes}`
}

export const getDateFromStringWithCurrentDateValue = (stringDate?: string): Date | undefined => {
  if (!stringDate) return

  const currentDate = new Date()
  const fullDate = new Date(stringDate)
  return new Date(
    currentDate.getFullYear(),
    currentDate.getMonth(),
    currentDate.getDate(),
    fullDate.getHours(),
    fullDate.getMinutes(),
    0,
    0
  )
}

export const stripDateFromSeconds = (date: Date): Date => {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), 0, 0)
}
