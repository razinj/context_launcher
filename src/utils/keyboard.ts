import { Keyboard } from 'react-native'

export const dismissKeyboard = () => Keyboard.dismiss()
