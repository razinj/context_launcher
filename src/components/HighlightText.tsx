import React from 'react'
import { StyleSheet, Text } from 'react-native'
import { useSelector } from 'react-redux'
import { PRIMARY_COLOR, WHITE_COLOR } from '../constants'
import { selectAppsSearchQuery } from '../slices/appState'
import { createKeyForHighlightTextElement } from '../utils/string'

type Props = {
  text: string
}

const HighlightText = ({ text }: Props) => {
  const searchQuery = useSelector(selectAppsSearchQuery)

  if (!searchQuery) {
    return <Text style={styles.normalText}>{text}</Text>
  }

  const regex = new RegExp(`(${searchQuery})`, 'gi')

  return (
    <Text style={styles.normalText}>
      {text.split(regex).map((substring: string, index: number) =>
        regex.test(substring) ? (
          <Text
            style={styles.highlightedText}
            key={createKeyForHighlightTextElement(index, substring)}
            testID='highlight-text-highlighted'>
            {substring}
          </Text>
        ) : (
          <Text key={createKeyForHighlightTextElement(index, substring)}>{substring}</Text>
        )
      )}
    </Text>
  )
}

const styles = StyleSheet.create({
  highlightedText: {
    color: PRIMARY_COLOR,
    fontWeight: '600',
    textShadowColor: 'rgba(255, 255, 255, 0.1)',
    textShadowRadius: 2.5,
  },
  normalText: {
    color: WHITE_COLOR,
    fontWeight: '600',
    textShadowColor: 'rgba(0, 0, 0, 0.75)',
    textShadowRadius: 2.5,
  },
})

export default HighlightText
