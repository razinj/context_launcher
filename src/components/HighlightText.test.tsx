import { screen } from '@testing-library/react-native'
import React from 'react'
import { initialStoreState } from '../../utils/test/data'
import { renderWithProvider } from '../../utils/test/utils'
import { RootState } from '../store'
import HighlightText from './HighlightText'

describe('<HighlightText /> Tests', () => {
  it('should render correctly and match snapshot - without search query', () => {
    const customInitialState: RootState = {
      ...initialStoreState,
      appState: {
        ...initialStoreState.appState,
        search: {
          query: undefined,
          result: [],
        },
      },
    }

    const textToHighlight = 'Chrome'

    renderWithProvider(<HighlightText text={textToHighlight} />, { preloadedState: customInitialState })

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByText(/Chrome/)).toBeOnTheScreen()
    expect(screen.queryAllByTestId('highlight-text-highlighted')).toHaveLength(0)
  })

  it('should render correctly and match snapshot - with search query', () => {
    const customInitialState: RootState = {
      ...initialStoreState,
      appState: {
        ...initialStoreState.appState,
        search: {
          query: 'chro',
          result: [], // left empty because result list doesn't matter in this component.
        },
      },
    }

    const textToHighlight = 'Chrome'

    renderWithProvider(<HighlightText text={textToHighlight} />, { preloadedState: customInitialState })

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByText(/Chrome/)).toBeOnTheScreen()
    expect(screen.getAllByTestId('highlight-text-highlighted')).toHaveLength(1)
  })

  it('should highlight query in text correctly', () => {
    const customInitialState: RootState = {
      ...initialStoreState,
      appState: {
        ...initialStoreState.appState,
        search: {
          query: 'o',
          result: [], // left empty because result list doesn't matter in this component.
        },
      },
    }

    const textToHighlight = 'Chrome & GOOGLE'

    renderWithProvider(<HighlightText text={textToHighlight} />, { preloadedState: customInitialState })

    expect(screen.getByText(/Chrome & GOOGLE/)).toBeOnTheScreen()

    const highlightedTexts = screen.getAllByTestId('highlight-text-highlighted')

    expect(highlightedTexts).toHaveLength(3)
    expect(highlightedTexts[0]).toHaveTextContent('o')
    expect(highlightedTexts[1]).toHaveTextContent('O')
    expect(highlightedTexts[2]).toHaveTextContent('O')
    // TODO: Check highlighted texts styling
  })
})
