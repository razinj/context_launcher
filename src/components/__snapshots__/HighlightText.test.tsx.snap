// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HighlightText /> Tests should render correctly and match snapshot - with search query 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontWeight": "600",
          "textShadowColor": "rgba(0, 0, 0, 0.75)",
          "textShadowRadius": 2.5,
        }
      }
    >
      <Text />
      <Text
        style={
          {
            "color": "#42855B",
            "fontWeight": "600",
            "textShadowColor": "rgba(255, 255, 255, 0.1)",
            "textShadowRadius": 2.5,
          }
        }
        testID="highlight-text-highlighted"
      >
        Chro
      </Text>
      <Text>
        me
      </Text>
    </Text>
  </View>
</RNCSafeAreaProvider>
`;

exports[`<HighlightText /> Tests should render correctly and match snapshot - without search query 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontWeight": "600",
          "textShadowColor": "rgba(0, 0, 0, 0.75)",
          "textShadowRadius": 2.5,
        }
      }
    >
      Chrome
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
